tensorflow/compiler/xla/service/llvm_ir/sort_util.cc

```diff
@@ -147,10 +147,11 @@ Status EmitTiledCompareLoop(
     const std::vector<IrArray>& params,
     const std::vector<llvm::Value*>& param_shmem_buffers, int64 tile_size,
     const EmitCallToNestedComputationCallback& emit_compare_callback,
-    llvm::IRBuilder<>* b) {
+    LLVMTargetIRBuilder& llvm_target_ir_builder) {
+  llvm::IRBuilder<>* b = llvm_target_ir_builder.builder();
   KernelSupportLibrary ksl(b);
-  llvm::Value* thread_id = llvm_ir::EmitCallToIntrinsic(
-      llvm::Intrinsic::nvvm_read_ptx_sreg_tid_x, {}, {}, b);
+  llvm::Value* thread_id = llvm_ir::EmitCallToTargetIntrinsic(
+                kTHREAD_ID_X, {}, {}, llvm_target_ir_builder);
```

Comment:
Actually I'm confused why we need to have this llvm_target_ir_builder at all.

Given an llvm::IRBuilder<>, can't we get the llvm::Target from the module?  And from that, can't we then translate kTHREAD_ID_X to the proper llvm::Intrinsic::ID?  It seems like that would greatly simplify things.

Reply:
Thanks for the suggestion, I have simplified the fix using the target triple to find the intrinsic. One thing I anticipate for the future is when one target needs a device function call  while other targets need intrinsics. Will need a way query  if the target needs a intrinsic vs a device function call. Would a query API like that reside in llvm_util ?

Reply:
Do callers actually need to *know* whether the thing that's being called is an LLVM intrinsic or a device function?  Could we just change the name of `EmitCallToTargetIntrinsic` to (say) `EmitCallToTargetSpecialFunction` and then handle the distinction between calling intrinsics and functions in there?

Reply:
Yeah..I think that might work.  I just find all the device function calls currently  handled separately and inside the emitters themselves. Might make sense to just have a generic  EmitCallToTargetSpecialFunction